src/Orleans/Storage/RelationalStorageExtensions.cs

```diff
@@ -0,0 +1,204 @@
+
+THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
+OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+using Orleans.Runtime.Storage.Relational;
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+using System.Threading.Tasks;
+
+
+namespace Orleans.Runtime.Storage.RelationalExtensions
+{
+    /// <summary>
+    /// Convenienience functions to work with objects of type <see cref="IRelationalStorage"/>.
+    /// </summary>
+    public static class RelationalStorageExtensions
+    {
+        /// <summary>
+        /// Used to format .NET objects suitable to relational database format.
+        /// </summary>
+        private static SqlFormatProvider sqlFormatProvider = new SqlFormatProvider();
+
+        /// <summary>
+        /// This is a template to produce query parameters that are indexed.
+        /// </summary>
+        private static string indexedParameterTemplate = "@p{0}";
+
+
+        /// <summary>
+        /// Executes a multi-record insert query clause with <em>SELECT UNION ALL</em>.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="storage">The storage to use.</param>
+        /// <param name="tableName">The table name to against which to execute the query.</param>
+        /// <param name="parameters">The parameters to insert.</param>
+        /// <param name="useSqlParams"><em>TRUE</em> if the query should be in parameterized form. <em>FALSE</em> otherwise.</param>
+        /// <returns>The rows affected.</returns>
+        public static Task<int> ExecuteMultipleInsertIntoAsync<T>(this IRelationalStorage storage, string tableName, IEnumerable<T> parameters, bool useSqlParams = false)
+        {
+            if(string.IsNullOrWhiteSpace(tableName))
+            {
+                throw new ArgumentException("The name must be a legal SQL table name", "tableName");
+            }
+
+            if(parameters == null)
+            {
+                throw new ArgumentNullException("parameters");
+            }
+
+            //SqlParameters map is needed in case the query needs to be parameterized in order to avoid two
+            //reflection passes as first a query needs to be constructed and after that when a database
+            //command object has been created, parameters need to be provided to them.
+            var sqlParameters = new Dictionary<string, object>();
+            const string insertIntoValuesTemplate = "INSERT INTO {0} ({1}) SELECT {2};";
+            string columns = string.Empty;
+            var values = new List<string>();
+            if(parameters.Any())
```

Comment:
Multiple enumerations of IEnumerable. Introduce variable for parameters.ToList() [or equivalent]

[Found by ReSharper]